src/librbd/api/Trash.cc

```diff
@@ -257,6 +257,11 @@ int Trash<I>::restore(librados::IoCtx &io_ctx, const std::string &image_id,
     return r;
   }
 
+  if (source_user && trash_spec.source == cls::rbd::TRASH_IMAGE_SOURCE_MIGRATION) {
```

Comment:
I would rather check for `trash_spec.source != cls::rbd::TRASH_IMAGE_SOURCE_USER`. Alternatively, instead of `bool source_user` we could pass `rbd_trash_image_source_t source` argument to `Trash<I>::restore` (similarly to `Trash<I>::move`) and proceed only if the sources match.